documentation/specs/multithreading/threading-fixes-and-task-migration-status.md

```diff
@@ -0,0 +1,558 @@
+
+**Description:** Improvements to how parameters are passed to TaskHost processes, including proper handling of execution context.
+
+**Changes Required:**
+
+#### src/Build/Instance/TaskFactories/TaskHostTask.cs
+- Ensure TaskEnvironment is properly serialized and passed to out-of-process TaskHost
+- Handle TaskHost lifecycle correctly in multithreaded scenarios
+
+**Rationale:** Tasks running in TaskHost processes need consistent environment regardless of execution mode.
+
+---
+
+### 11. Additional Concurrent Collections
+
+**Status:** Some already in main, others may be needed
+
+**Description:** Various internal data structures may need to be converted to thread-safe collections for multithreaded execution.
+
+**Known Conversions:**
+- `NodeProviderOutOfProcTaskHost._nodeIdToPacketFactory` - **Already in main** (changed to `ConcurrentDictionary`)
+- `NodeProviderOutOfProcTaskHost._nodeIdToPacketHandler` - **Already in main** (changed to `ConcurrentDictionary`)
+
+**May Still Need Review:**
+- Other internal caches and registries that are accessed during concurrent task execution
+- Task factory registries
+- Logger state management
+
+**Rationale:** Without thread-safe collections, concurrent access can cause race conditions, data corruption, or crashes in multithreaded builds.
+
+---
+
+## Tasks Migration to New API Interface
+
+The following section documents which tasks have been migrated to implement `IMultiThreadableTask` in the feature branch but are still using legacy execution in main.
+
+### Migration Status
+
+The feature branch has migrated **25 tasks** to implement `IMultiThreadableTask`. These tasks use `TaskEnvironment` for thread-safe access to working directory and environment variables.
+
+### Tasks Migrated in Feature Branch (Not in Main)
+
+#### File I/O Tasks
+1. **Copy** (`src/Tasks/Copy.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment.GetAbsolutePath()` for source and destination paths
+   - Uses `TaskEnvironment.GetEnvironmentVariable()` for configuration checks
+   - Passes `TaskEnvironment` to `FileState` constructors
+
+2. **Delete** (`src/Tasks/Delete.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for path resolution
+
+3. **MakeDir** (`src/Tasks/MakeDir.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for directory path resolution
+
+4. **RemoveDir** (`src/Tasks/RemoveDir.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for directory operations
+
+5. **Touch** (`src/Tasks/Touch.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for file path resolution
+
+6. **ReadLinesFromFile** (`src/Tasks/FileIO/ReadLinesFromFile.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for file access
+
+7. **WriteLinesToFile** (`src/Tasks/FileIO/WriteLinesToFile.cs`)
+   - Implements `IMultiThreadableTask`
+   - Uses `TaskEnvironment` for file writing
```

Comment:
create changes on this branch for file iO task migration. Note that you need both the attribute and the interface to work on main